infra/stream/Formatter.cpp

```diff
@@ -0,0 +1,225 @@
+        const auto size = std::numeric_limits<uint64_t>::digits + 2; // dot and sign
+        char buffer[size];
+        auto ptr = &buffer[size - 1];
+
+        decltype(value) divider = 10;
+        switch (spec.type)
+        {
+        case 'x':
+        case 'X':
+            divider = 16;
+            break;
+        case 'o':
+            divider = 8;
+            break;
+        case 'b':
+            divider = 2;
+            break;
+        default:
+            break;
+        }
+        const auto digits = isupper(spec.type) ? "0123456789ABCDEF" : "0123456789abcdef";
+        for (; value >= divider; value /= divider)
+        {
+            *ptr-- = digits[value % divider];
+            if (--dotPos == 0)
+                *ptr-- = '.';
+        }
+        *ptr = digits[value];
+        if (negative)
+            *--ptr = '-';
+        if (spec.align == FormatAlign::nothing)
+            spec.align = FormatAlign::right;
+        RawFormat(writer, ptr, &buffer[size - 1] - ptr + 1, spec);
+    }
+
+    void FormatterBase::SignedInteger(int32_t value, StreamWriter& writer, FormatSpec& spec)
+    {
+        const auto negative = value<0;
+        RawInteger(negative ? -value : value, negative, 0, writer, spec);
+    }
+
+    void FormatterBase::UnsignedInteger(uint32_t value, StreamWriter& writer, FormatSpec& spec)
+    {
+        RawInteger(value, false, 0, writer, spec);
+    }
+
+    template<>
+    void Formatter<int32_t>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        SignedInteger(value, writer, spec);
+    }
+
+    template<>
+    void Formatter<int16_t>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        SignedInteger(value, writer, spec);
+    }
+
+    template<>
+    void Formatter<int8_t>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        SignedInteger(value, writer, spec);
+    }
+
+    template<>
+    void Formatter<uint8_t>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        UnsignedInteger(value, writer, spec);
+    }
+
+    template<>
+    void Formatter<uint16_t>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        UnsignedInteger(value, writer, spec);
+    }
+
+    template<>
+    void Formatter<char const*>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        RawFormat(writer, value, strlen(value), spec);
+    }
+
+    template<>
+    void Formatter<bool>::Format(StreamWriter& writer, FormatSpec& spec)
+    {
+        const auto str = value ? "true" : "false";
+        RawFormat(writer, str, strlen(str), spec);
+    }
+
+    FormatWorker::FormatWorker(StreamWriter& writer, const char* formatStr, std::vector<FormatterBase*> &formatters)
+            : format(formatStr)
+    {
+        StreamErrorPolicy errorPolicy;
+        for (;;)
+        {
+            if (IsEndFormat())
+                return;
+            const auto ch = *format++;
+            if (ch != '{')
+            {
+                writer.Insert(MakeByteRange(ch), errorPolicy);
```

Comment:
Funky indentation over here. Either this line is indented too much, or the next one too little.